### PR TITLE
meta/autoid: add TestAsyn

### PR DESCRIPTION
**Pull request by zimulala**
meta/autoid: add TestAsyn

### CODE REVIEW

File: meta/autoid/autoid_test.go [hunk 1]
```diff
@@ -120,3 +122,69 @@ func (*testSuite) TestT(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(id, Equals, int64(6544))
 }
+
+func (*testSuite) TestAsyn(c *C) {
+	driver := localstore.Driver{Driver: goleveldb.MemoryDriver{}}
+	store, err := driver.Open("memory")
+	c.Assert(err, IsNil)
+	defer store.Close()
```
Comment (hanfei1991) [inline meta/autoid/autoid_test.go]: can we put this defer to line 133?

File: meta/autoid/autoid_test.go [hunk 2]
```diff
@@ -120,3 +122,69 @@ func (*testSuite) TestT(c *C) {
...
+		err = m.CreateTable(dbID, &model.TableInfo{ID: tblID, Name: model.NewCIStr("t")})
+		c.Assert(err, IsNil)
+		return nil
+	})
+	c.Assert(err, IsNil)
+
+	var mu sync.Mutex
+	wg := sync.WaitGroup{}
+	m := map[int64]struct{}{}
+	errCh := make(chan error, 100)
```
Comment (hanfei1991) [inline meta/autoid/autoid_test.go]: make 100 a constant?

File: meta/autoid/autoid_test.go [hunk 3]
```diff
@@ -120,3 +122,69 @@ func (*testSuite) TestT(c *C) {
...
+
+	allocIDs := func() {
+		alloc := NewAllocator(store, dbID)
```
Comment (hanfei1991) [inline meta/autoid/autoid_test.go]: how about add defer wg.Done() here?

File: meta/autoid/autoid_test.go [hunk 4]
```diff
@@ -120,3 +122,69 @@ func (*testSuite) TestT(c *C) {
...
+		time.Sleep(time.Duration(i%10) * time.Microsecond)
+
+		go func() {
+			defer wg.Done()
+			allocIDs()
+		}()
+	}
+	wg.Wait()
+
+	if len(errCh) > 0 {
```
Comment (hanfei1991) [inline meta/autoid/autoid_test.go]: close the channel and err := <- errCh directly?

File: meta/autoid/autoid_test.go [hunk 5]
```diff
@@ -120,3 +122,69 @@ func (*testSuite) TestT(c *C) {
...
+				mu.Unlock()
+				break
+			}
+			m[id] = struct{}{}
+			mu.Unlock()
+		}
+	}
+	for i := 0; i < 100; i++ {
+		wg.Add(1)
+		time.Sleep(time.Duration(i%10) * time.Microsecond)
```
Comment (hanfei1991) [inline meta/autoid/autoid_test.go]: Why sleep here ?

File: meta/autoid/autoid_test.go (hunk 1 above)
Comment (zimulala) [inline meta/autoid/autoid_test.go]: `defer store.Close()` under the operation of opening  is clearer. And it's in test that doesn't need to consider performance.

File: meta/autoid/autoid_test.go (hunk 5 above)
Comment (zimulala) [inline meta/autoid/autoid_test.go]: Make some allocators start at the same time, some allocators don't start at the same time.

File: meta/autoid/autoid_test.go [hunk 6]
```diff
@@ -120,3 +122,69 @@ func (*testSuite) TestT(c *C) {
...
+			id, err := alloc.Alloc(tblID)
+			if err != nil {
+				errCh <- err
+				break
+			}
+
+			mu.Lock()
+			if _, ok := m[id]; ok {
+				errCh <- fmt.Errorf("duplicate id:%v", id)
+				mu.Unlock()
```
Comment (hanfei1991) [inline meta/autoid/autoid_test.go]: how about defer mu.Unlock()?

File: meta/autoid/autoid_test.go [hunk 7]
```diff
@@ -120,3 +122,69 @@ func (*testSuite) TestT(c *C) {
...
+		time.Sleep(time.Duration(i%10) * time.Microsecond)
+
+		go func() {
+			defer wg.Done()
+			allocIDs()
+		}()
+	}
+	wg.Wait()
+
+	select {
```
Comment (hanfei1991) [inline meta/autoid/autoid_test.go]: How about check if len(errCh) is equal to 0 ?

File: meta/autoid/autoid_test.go (hunk 6 above)
Comment (zimulala) [inline meta/autoid/autoid_test.go]: It think is OK

File: meta/autoid/autoid_test.go [hunk 8]
```diff
@@ -120,3 +122,68 @@ func (*testSuite) TestT(c *C) {
...
+			}
+			m[id] = struct{}{}
+			mu.Unlock()
+		}
+	}
+	for i := 0; i < count; i++ {
+		wg.Add(1)
+		go func(num int) {
+			time.Sleep(time.Duration(num%10) * time.Microsecond)
+			defer wg.Done()
```
Comment (hanfei1991) [inline meta/autoid/autoid_test.go]: use wg.Done below line 179 directly.

File: meta/autoid/autoid_test.go [hunk 9]
```diff
@@ -120,3 +122,65 @@ func (*testSuite) TestT(c *C) {
...
+		go func(num int) {
+			defer wg.Done()
+			time.Sleep(time.Duration(num%10) * time.Microsecond)
+			allocIDs()
+		}(i)
+	}
+	wg.Wait()
+
+	close(errCh)
+	err, ok := <-errCh
```
Comment (coocood) [inline meta/autoid/autoid_test.go]: Why not just receive err, and check if err is nil?

File: meta/autoid/autoid_test.go [hunk 10]
```diff
@@ -120,3 +122,65 @@ func (*testSuite) TestT(c *C) {
 	c.Assert(err, IsNil)
 	c.Assert(id, Equals, int64(6544))
 }
+
+func (*testSuite) TestAsyn(c *C) {
```
Comment (coocood) [inline meta/autoid/autoid_test.go]: `Asyn` doesn't properly describe the test.
And add comment about the purpose of the test.

**Comment by coocood** [comment]
LGTM

**Comment by zimulala** [comment]
PTAL @hanfei1991 @tiancaiamao

**Comment by hanfei1991** [comment]
LGTM

**Pull request merged**